src/cryptoadvance/specter/user.py

```diff
@@ -36,69 +44,146 @@ def verify_password(stored_password, provided_password):
-                    user_dict["password"],
-                    user_dict["config"],
-                    specter,
-                )
-            else:
-                return cls(
-                    user_dict["id"],
-                    user_dict["username"],
-                    user_dict["password"],
-                    {},
-                    specter,
-                    is_admin=True,
+        # create cipher config
+        cipher_config = AES.new(private_key, AES.MODE_GCM)
+
+        # return a dictionary with the encrypted text
+        cipher_text, tag = cipher_config.encrypt_and_digest(self.plaintext_user_secret)
+        self.encrypted_user_secret = {
+            "cipher_text": b64encode(cipher_text).decode("utf-8"),
+            "salt": b64encode(salt).decode("utf-8"),
+            "nonce": b64encode(cipher_config.nonce).decode("utf-8"),
+            "tag": b64encode(tag).decode("utf-8"),
+        }
+
+    def decrypt_user_secret(self, plaintext_password):
+        # See: https://qvault.io/cryptography/aes-256-cipher-python-cryptography-examples/
+        if not self.encrypted_user_secret:
+            self._generate_user_secret(plaintext_password)
+
+        # decode the dictionary entries from base64
+        salt = b64decode(self.encrypted_user_secret["salt"])
+        cipher_text = b64decode(self.encrypted_user_secret["cipher_text"])
+        nonce = b64decode(self.encrypted_user_secret["nonce"])
+        tag = b64decode(self.encrypted_user_secret["tag"])
+
+        # generate the private key from the password and salt
+        private_key = hashlib.scrypt(
+            plaintext_password.encode(), salt=salt, n=2 ** 14, r=8, p=1, dklen=32
+        )
+
+        # create the cipher config
+        cipher = AES.new(private_key, AES.MODE_GCM, nonce=nonce)
+
+        # decrypt the cipher text
+        self.plaintext_user_secret = cipher.decrypt_and_verify(cipher_text, tag)
+
+    def _generate_user_secret(self, plaintext_password):
+        # Encryption using the user_secret uses a Fernet key. But the Fernet
+        #   key itself will be encrypted with the user's password.
+        self.plaintext_user_secret = Fernet.generate_key()
+        self._encrypt_user_secret(plaintext_password)
+        self.save_info()
+
+    def set_password(self, plaintext_password):
+        # Hash the incoming plaintext password and update the encrypted
+        #   user_secret as needed.
+        self.password_hash = hash_password(plaintext_password)
+
+        # Must keep encrypted_user_secret in sync with password changes
+        if self.encrypted_user_secret is None:
+            self._generate_user_secret(plaintext_password)
+        else:
+            if self.plaintext_user_secret is None:
+                raise Exception(
+                    "encrypted_user_secret wasn't decrypted during user login"
                 )
-        except Exception as e:
-            raise SpecterError(f"Unable to parse user JSON.:{e}")
```

Comment:
please `handle_exception(e)` whenever raising Exceptions (or be more specific in the exception you're catching).
We never know what kind of exception will come up here and if it's an unexpected one, we definitely want to have a stacktrace (and handle_exception will print one out).